manifests/init.pp

```diff
@@ -5,27 +5,112 @@
 # SIMP provides a utility for exactly that, `/usr/local/bin/unpack_dvd`
 #
-# @author Trevor Vaughan <tvaughan@onyxpoint.com>
+# @param package_list List of packages related to libvirt to be managed
+# @param service_ensure Ensure setting for libvirtd
+# @param ksm Enable management of ksm
+# @param kvm Enable management of kvm
+# @param load_kernel_modules Enable management of kernel modules in this module
 #
-class libvirt {
+# @param ksm_max_kernel_pages
+#   The maximum number of unswappable kernel pages which may be allocated by
+#   ksm (0 for unlimited) If unset, defaults to half of total memory.
+#
+# @param ksm_monitor_interval
+#   The number of seconds ksmtuned should sleep between tuning adjustments
+#   Every KSM_MONITOR_INTERVAL seconds ksmtuned adjust how aggressive KSM
+#   will search for duplicated pages based on free memory.
+#
+# @param ksm_sleep_msec
+#   Millisecond sleep between ksm scans for 16Gb server.  Smaller servers
+#   sleep more, bigger sleep less.  How many Milliseconds to sleep between
+#   scans of 16GB of RAM.  The actual sleep time is calculated as sleep =
+#   KSM_SLEEP_MSEC * 16 / Total GB of RAM The final sleep value will be
+#   written to /sys/kernel/mm/ksm/sleep_millisecs
+#
+# @param ksm_npages_boost
+#   Amount to increment the number of pages to scan. The number of pages to
+#   be scanned will be increased by KSM_NPAGES_BOOST when the amount of free
+#   ram < threshold (see KSM_THRES_* below)
+#
+# @param ksm_npages_decay
+#   Amount to decrease the number of pages to scan The number of pages to be
+#   scanned will be decreased by KSM_NPAGES_DECAY when the amount of free ram
+#   >= threshold (see KSM_THRES_* below)
+#
+# @param ksm_npages_min
+#   Minimum number of pages to be scanned at all times
+#   If this variable is set to 'shmall', then half of the value in
+#   /proc/sys/kernel/shmall will be used.
+#
+# @param ksm_npages_max
+#   Maximum number of pages to be scanned at all times
+#   If this variable is set to 'shmall', then the value in
+#   /proc/sys/kernel/shmall will be used.
+#
+# @param ksm_thres_coef
+#   If free memory is less than this percentage KSM will be activated.
+#   NOTE: Only KSM_THRES_CONST or KSM_THRES_COEF is actually used.  Whichever
+#   results in a larger number wins.
+#
+# @param ksm_thres_const
+#   If free memory is less than this number KSM will be activated
+#   NOTE: Only KSM_THRES_CONST or KSM_THRES_COEF is actually used.  Whichever
+#   results in a larger number wins.
+#
+# @param kvm_package_list List of packages related to kvm to be managed
+# @param manage_sysctl Disable to avoid setting sysctl settings in this module.
+# @param package_ensure Ensure setting for all packages in this module
+#
+# @author https://github.com/simp/pupmod-simp-libvirt/graphs/contributors
+#
+class libvirt (
+  Array[String] $package_list,
+  String $service_ensure,
+  Boolean $ksm,
+  Boolean $kvm,
+  Boolean $load_kernel_modules,
+
+  # KSM
```

Comment:
KSM is not uniquely bound to `libvirt` but is included by `qemu-kvm`. This should not be tied into the top level class at all.

Reply:
including `libvirt::ksm` eventually includes `libvirt` as well, so it's that's the case it should be fully decoupled

Reply:
but I didn't know that

Reply:
@jeefberkey Yeah, KSM should probably be broken out into another module and we should use 'ensure_resource' on all of the packages. Unfortunately, that's going to need to be a SIMP 7 change since it would be breaking.